[CORE-12556] Add a scale test for Topic ID Migration

BenPope opened the pull request:
The test showed up an oversize allocation, which is fixed.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

michael-redpanda committed:
dt/scale: Added topic_id migration scale test

This scale test will validate that the topic_id migrator, active on
v25.2 release, will scale.  The test will verify that it works against
10, 100, 1000, and 10_000 topics after upgrading from 25.1 to 25.2.

Signed-off-by: Michael Boquard <michael@redpanda.com>

BenPope committed:
treewide: Fix includes

The next commit will break a dependency between serde and future-util.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
serde: Do not depend on future-util

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
cluster: Parameterize Container for make_error_topic_results

Allow specifying output container

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
cluster: Use chunked_vector for topic prop update

Signed-off-by: Ben Pope <ben@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--50e76acf-dd2a-45da-bc56-ddbff4bff5a1-->
<!--TEST RESULTS-->
<!--3913110c-fae2-434b-a8d6-2a083cf5f202-->
<!--TEST RESULTS-->
<!--4e41f390-3a8a-454a-8c33-66a82834e8d1-->
<!--TEST RESULTS-->
<!--2d14143f-e3c7-4595-b512-8044af07fa74-->
<!--TEST RESULTS-->
<!--73edeaa9-e022-4970-8953-de2fa853d781-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68963">build#68963</a></summary>


| test_class            | test_method                               | test_arguments                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:----------------------|:------------------------------------------|:---------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest          | test_timeboxed_uploads                    | {"acks": 0, "cloud_storage_type": 2}                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68963#01980b5f-19d6-428f-bd6b-1aea0da72b82 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
| MaintenanceTest       | test_maintenance_sticky                   | {"use_rpk": false}                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68963#01980b5f-19d1-4074-b0b5-39f2ba8fbd59 | FLAKY         | 19/21    | upstream reliability is '98.94459102902374'. current run reliability is '90.47619047619048'. drift is 8.4684 and the allowed drift is set to 50. The test should PASS |
| NodePoolMigrationTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68963#01980b5f-19d5-416f-9f8d-2206bf1e5e29 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68993">build#68993</a></summary>


| test_class       | test_method                 | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-----------------|:----------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests | test_latest_protobuf_schema | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68993#01980da3-cec0-478b-b869-4b8577d7e4bf | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69078">build#69078</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69078#01981226-3b0a-43bb-a30b-5d9bd493b8e4 | FLAKY         | 19/21    | upstream reliability is '99.29328621908127'. current run reliability is '90.47619047619048'. drift is 8.8171 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69109">build#69109</a></summary>


| test_class      | test_method             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------|:------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudTopicsTest | test_reconciler_uploads | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69109#0198144a-1c4b-4514-a535-3767db7c7559 | FLAKY         | 20/21    | upstream reliability is '99.78991596638656'. current run reliability is '95.23809523809523'. drift is 4.55182 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69164">build#69164</a></summary>


| test_class                          | test_method         | test_arguments             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:------------------------------------|:--------------------|:---------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| cluster_metadata_uploader_fixture   | test_upload_in_term |                            | unit        | https://buildkite.com/redpanda/redpanda/builds/69164#019817de-8f75-4151-bf0a-aee46dde8e68 | FAIL          | 0/1      |                                                                                                                                                           |
| NodeDecommissionSpaceManagementTest | test_decommission   | {"single_partition": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69164#01981815-9344-494a-bab5-d54172a65a75 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

IoannisRP commented on src/v/ssx/future-util.h:
```diff
@@ -241,7 +250,10 @@ inline auto parallel_transform(Iterator begin, Iterator end, Func func) {
 /// of the exceptions.
-template<typename Rng, typename Func>
+template<
+  typename Rng,
+  typename Func,
+  template<typename> class Container = detail::unary_std_vector>
```
if you put `Container` as the first argument, you will be able to define the return type as the single parameter argument without needed to pass the other 2. i.e. `parallel_transform<chunked_vector>(...)`

BenPope replied:
> if you put `Container` as the first argument, you will be able to define the return type as the single parameter argument without needed to pass the other 2. i.e. `parallel_transform<chunked_vector>(...)`

But then it can't be defaulted. I suppose I could fix up the 24 call sites, or restrict an overload set with concepts. I'll look into it.

IoannisRP replied:
```
template<
  template<typename> class Container = detail::unary_std_vector,
  typename Rng,
  typename Func>
requires requires(Func f, Rng r) {
```
this works just fine, doesn't it? When you call it, it can deduct all 3 types (1 defaulted, 2 from args) and it doesn't complain

BenPope replied:
Not sure why I thought that didn't work, thanks.

IoannisRP commented on src/v/ssx/future-util.h:
```diff
@@ -250,13 +262,47 @@ requires requires(Func f, Rng r) {
+    r.end();
+    { r.begin() != r.begin() } -> std::convertible_to<bool>;
+}
+inline auto parallel_transform_fragmented(Rng rng, Func func) {
+    return parallel_transform<Rng, Func, Container>(
+      std::move(rng), std::move(func));
+}
```
a spot test would be nice. If nothing else, for documentation purposes.

IoannisRP commented on src/v/cluster/cluster_utils.h:
```diff
@@ -30,13 +31,16 @@
+template<
+  std::ranges::sized_range Rng,
+  typename Fn,
+  template<typename> class Container = ssx::detail::unary_std_vector>
```
The fact that you use it outside the header means it's probably not a `detail`.

BenPope replied:
moved to a namespace

IoannisRP reviewed (commented):
The untangling of the header/build deps is just :fire: !

Given the size of the changes and that you add a new util function in ssx (for visibility), it's probably worth the extra effort to split the first 3 commits into a separate pr.

BenPope commented:
> Given the size of the changes and that you add a new util function in ssx (for visibility), it's probably worth the extra effort to split the first 3 commits into a separate pr.

Ideally `future-util.h` should be split up, too.

IoannisRP commented on src/v/ssx/future-util.h:
```diff
@@ -28,6 +28,9 @@
 #include <memory>
 #include <type_traits>
 
+template<typename T, size_t fragment_size_bytes>
+class fragmented_vector;
+
```
what is the reason for the forward declare here instead of adding the header?

BenPope replied:
> what is the reason for the forward declare here instead of adding the header?

So that I don't have to depend on `container/fragmented_vector`

IoannisRP replied:
But it kinda does.
- There is a typedef to a `fragmented_vector`
- It is the default for a function

It's not a big issue, as most places will have the header included. But when the error comes because of the missing header, it is a weird one. I think that adding the header is worth the cost.

BenPope replied:
Now included.

IoannisRP commented on src/v/ssx/future-util.h:
```diff
@@ -62,6 +65,17 @@ async_transform(Iterator begin, Iterator end, Func&& func) {
+using std_vector = std::vector<T>;
+
+template<typename T>
+using fragmented_vector = fragmented_vector<T, 8192>;
```
would `chunked_vector` be a better default?

BenPope replied:
Changed

IoannisRP commented on src/v/ssx/future-util.h:
```diff
@@ -217,11 +234,8 @@ inline auto parallel_transform(Iterator begin, Iterator end, Func func) {
-          std::make_move_iterator(res.begin()),
-          std::make_move_iterator(res.end()));
-    });
+    return ssx::when_all_succeed<Container<typename future::value_type>>(
+      std::move(res));
```
No-action-comment:
i like what you did with the templated `Container`. `when_all_succeed` could be updated in the same way.
the unary_container helper could be moved out and reused across `ssx`.

BenPope replied:
Since I'm now using `chunked_vector`, which doesn't have any non-type template arguments, I don't need it any more.

IoannisRP commented on tests/rptest/scale_tests/topic_id_migration.py:
```diff
@@ -0,0 +1,144 @@
+            self.logger.debug(
+                f"{str(node.account)}: Could not parse as json: {str(jsondata)}"
+            )
+            return None
```
if you return an empty dict on failure, it saves you an if on the calling site.

IoannisRP commented on tests/rptest/scale_tests/topic_id_migration.py:
```diff
@@ -0,0 +1,144 @@
+        self.logger.debug(f"...cmd: {cmd}")
+
+        data = {}
+        for line in node.account.ssh_capture(cmd):
+            self.logger.debug(
+                f"received {sys.getsizeof(line)}B from '{hostname}'.")
+            data = self._try_parse_json(node, line.strip())
+            if data is not None:
+                if 'error' in data:
+                    self.logger.warning(
+                        f"Node '{hostname}' reported error:\n{data['error']}")
+                    raise RuntimeError(data['error'])
+            else:
+                data = {}
```
how does this loop work?

We overwrite data on every iteration. For all the other ones, we are just doing error checking?

BenPope replied:
@michael-redpanda ?

IoannisRP reviewed (dismissed):
Overall looking ready. Just a couple of questions

BenPope committed:
ssx: Parameterize Container for parallel_transform

Allow transforming a large container whilst avoiding an oversize
allocation.

The switch to ssx::when_all_succeed is necessary to avoid the use of
std::vector in ss::when_all_succeed.

Signed-off-by: Ben Pope <ben@redpanda.com>

IoannisRP commented on src/v/ssx/future-util.h:
```diff
@@ -250,13 +254,43 @@ requires requires(Func f, Rng r) {
+/// function invocations that resolves when all the function invocations
+/// complete.  If one or more return an exception, the return value contains one
+/// of the exceptions.
+template<
+  template<typename...> class Container = chunked_vector,
+  typename Rng,
+  typename Func>
+requires requires(Func f, Rng r) {
+    r.begin();
+    r.end();
+    { r.begin() != r.begin() } -> std::convertible_to<bool>;
+}
+inline auto parallel_transform_chunked(Rng rng, Func func) {
+    return parallel_transform<Container>(std::move(rng), std::move(func));
+}
```
nit: 
```suggestion
template<
  typename Rng,
  typename Func>
requires requires(Func f, Rng r) {
    r.begin();
    r.end();
    { r.begin() != r.begin() } -> std::convertible_to<bool>;
}
inline auto parallel_transform_chunked(Rng rng, Func func) {
    return parallel_transform<chunked_vector>(std::move(rng), std::move(func));
}
```
given that we are just forwarding the call to `parallel_transform` there is no benefit to being able to call
`parallel_transform_chunked<my_vector>(...)`. You can just call `parallel_transfrom<my_vector>(...)`

BenPope replied:
> You can just call parallel_transfrom<my_vector>(...)

I'll just remove it, but keep the test.

IoannisRP replied:
i didn't mind the extra function doing the specialization, as it is what we are inclined to use. I just meant that the specialization doesn't need to have a customization point. Anyway, it's good either way.

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/266477020aa0297df361efa75e9a65dd983b1e8e..e62d33559672ac5a3ed981fe914715c3c812b53e)
* Rebased due to missing include in new code (`cluster_link/link.cc`)
* `future-util.h`: Removed `parallel_transform_chunked` -> Just call `parallel_transform<chunked_vector>`
* `topic_id_migration_test.pt`: Returned empty dict to avoid checking against None

BenPope commented on src/v/ssx/BUILD:
```diff
@@ -36,6 +36,8 @@ redpanda_cc_library(
     visibility = ["//visibility:public"],
     deps = [
         "//src/v/base",
+        "//src/v/container:fragmented_vector",
```
This is unused

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/07989cd7a15be7e585f9b513f10262cd0569fdfb..c06af9a2fc1fae3f9c27f6d9f7020f544bf9c303)
* Rebase
* Fixup new includes

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-26813-v25.2.x-350 remotes/upstream/v25.2.x
git cherry-pick -x 75565d4712 fc44693a09 9fc11344a5 30a9e0e31e c5c88518c5 c06af9a2fc
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/16347585893).
